### PR TITLE
[SYSTEMDS-3758] Python API Builtin triu, tril, argmin, argmax and casting Scalar <-> Matrix <-> Frame

### PR DESCRIPTION
**Pull request by e-strauss**
This patch adds the builtin functions tril, triu, argmin (rowIndexMin) and argmax (rowIndexMax) to the python API. Additionally, I added functionalities to the python API to 
 - cast matrices, frames and scalars 
 - reshape matrices

I also extended the Scalar Output Type with boolean and integer

### CODE REVIEW

File: src/main/python/systemds/operator/nodes/matrix.py [hunk 1]
```diff
@@ -634,6 +634,14 @@ def to_string(self, **kwargs: Dict[str, VALID_INPUT_TYPES]) -> "Scalar":
             self.sds_context, "toString", [self], kwargs, output_type=OutputType.STRING
         )
 
+    def to_scalar(self):
+        return Scalar(self.sds_context, "as.scalar", [self])
+
+    def to_frame(self):
+        from systemds.operator import Frame
```
Comment (Baunsgaard) [inline src/main/python/systemds/operator/nodes/matrix.py:641]: cyclic dependency, i understand it is how i thought about it as well, however, maybe we need to add it afterwards dynamically instead.

See:

https://stackoverflow.com/questions/972/adding-a-method-to-an-existing-object-instance-in-python

This would allow us to add the to_scalar, to_frame and to_matrix on the operation_node directly inside frame, matrix and scalar themselves to overwrite the operation node.

This would also allow us to have documentation on the method, since you add an empty method with docs that would throw an exception on the operation_node, and then on the import of Matrix etc it overwrite the Class method such that new instances use the overwritten method in Matrix.

File: src/main/python/systemds/operator/nodes/matrix.py [hunk 2]
```diff
@@ -810,5 +818,89 @@ def quantile(self, p, weights: "Matrix" = None) -> "OperationNode":
...
+                f"Axis has to be either 0, 1 or None, for column, row or complete {self.operation}"
+            )
+
+    def reshape(self, rows, cols=1):
+        """Gives a new shape to a matrix without changing its data.
+
+        :param rows: number of rows
+        :param cols: number of columns, defaults to 1
+        :return: `Matrix` representing operation"""
+        return Matrix(self.sds_context, "matrix", [self, rows, cols])
```
Comment (Baunsgaard) [inline src/main/python/systemds/operator/nodes/matrix.py:903]: the rest of the functions looks good.

File: src/main/python/systemds/operator/nodes/scalar.py
Comment (Baunsgaard) [inline src/main/python/systemds/operator/nodes/scalar.py:1]: please move the formatting to another PR

File: src/main/python/systemds/script_building/dag.py
Comment (Baunsgaard) [inline src/main/python/systemds/script_building/dag.py:1]: same here, move the formatting to another PR.

File: src/main/python/systemds/script_building/dag.py [hunk 3]
```diff
@@ -43,6 +43,8 @@ class OutputType(Enum):
     NONE = auto()
     SCALAR = auto()
     STRING = auto()
+    INT = auto()
+    BOOLEAN = auto()
```
Comment (Baunsgaard) [inline src/main/python/systemds/script_building/dag.py:47]: These are not valid, since both are instances of SCALAR,
where are they used?

**Review by Baunsgaard — COMMENTED** [review]
Most of the content is good, however i really hope the casting can be modified according to my comment, i am not 100% sure it is possible but it would be great.

Furthermore, the PR contains formatting of frame and scalar, that should be moved to another PR.

File: src/main/python/systemds/operator/nodes/frame.py [hunk 4]
```diff
@@ -112,68 +136,94 @@ def transform_apply(self, spec: "Scalar", meta: "Frame"):
...
+
+    def to_scalar(self):
+        return Scalar(self.sds_context, "as.scalar", [self])
 
     def __str__(self):
         return "FrameNode"
 
-    def nRow(self) -> 'Scalar':
-        return Scalar(self.sds_context, 'nrow', [self])
+    def nRow(self) -> "Scalar":
```
[the remark this answers is not in the document]
Comment (e-strauss) [reply src/main/python/systemds/operator/nodes/frame.py:190]: All the formatting comes from black, as you recommended

File: src/main/python/systemds/script_building/dag.py (hunk 3 above)
Comment (e-strauss) [reply src/main/python/systemds/script_building/dag.py:47]: But isnt STRING also a value type of Scalar?

I added boolean and int there since in the scalar class, the value type of the scalar is determined by this. But you're right it would make more sense to put the value type in a different enum and create a new attribute, so that there is one field for the data type (matrix, frame, scalar) and one for the value types (double, integer, string, boolean)

File: src/main/python/systemds/operator/nodes/frame.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Baunsgaard) [reply src/main/python/systemds/operator/nodes/frame.py:190]: okay, maybe we should add a verification in Github actions that analyze if the code base is all formated according to black, i have seen this in other repositories such as Modyn

The check would be simple to add in a new action using: 
https://black.readthedocs.io/en/stable/usage_and_configuration/the_basics.html#check

File: src/main/python/systemds/script_building/dag.py (hunk 3 above)
Comment (Baunsgaard) [reply src/main/python/systemds/script_building/dag.py:47]: you are right, that would make more sense.

I think the reason why it is as it is now, is because someone (probably me) was lazy when making the initial version.

File: src/main/python/systemds/operator/nodes/matrix.py (hunk 1 above)
Comment (e-strauss) [reply src/main/python/systemds/operator/nodes/matrix.py:641]: I see, currently the cyclic dependency error is avoided since the import happens inside the function, which won't be executed while importing afaik.

File: src/main/python/systemds/script_building/dag.py (hunk 3 above)
Comment (e-strauss) [reply src/main/python/systemds/script_building/dag.py:47]: I will continue on this PR once the other Datatype / Valuetype PR is gone through

**Comment by e-strauss** [comment]
@Baunsgaard I adapted the casting, please take a look

**Comment by Baunsgaard** [comment]
LGTM
